storage: principled cross-collection since handling

aljoscha opened the pull request:
### Motivation

Modern-day revival of #13504.

This is updated to make the dependency tracking of sinks more principaled. And will hopefully also help simplify how we track the dependency on the remap shard in #15813 .

Before, we were tracking for each collection any sinks/exports that
depended on that collection. And we had somewhat complicated code that
was making sure we take their requirements into account when calculating
a new since for a collection.

Now, we turn that relationship around. Exports register requirements on
the since of their dependencies (the collection that they are sinking
from). And we track those dependencies and forward any changes in
requirements. Similar how the compute controller tracks since changes
between collections (including storage collections).

### Tips for reviewer

I left some TODOs/concerns in the code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -1299,30 +1347,58 @@ where
+            // would not have compacted up to `since`! This is tested by the
+            // kafka persistence tests.
+            //
+            // Aljoscha: I think we don't need to step back here. The previous
```
:skull: Take special care when reading this!

bkirwi replied:
Ugh. That sounds right, but I'll need to rummage around a bit to feel more confident.

Thinking about the equivalent situation in compute: a trivial MV with a single input collection. The frontier of that output will advance all the way to the frontier of the input, right? (And the only reason we can't do that for a sink is the "snapshot problem" - that we can't distinguish between updates from immediately before the as-of and updates that happened just after.)

bkirwi replied:
Now rather more confident, though I'm still interested in your thought on the MV question!

aljoscha replied:
@petrosagg I agree! Sinks are also weird in the frontier they write out. I think they will still need to write out the `F-1` frontier, because that's what they currently do and we can't easily change it now.

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -1076,10 +1105,13 @@ where
                     )
                     .await;
 
+                let storage_dependencies = this.get_collection_dependencies(&description);
```
Discussed this in person, but it seems like these dependencies also need to be fed into `install_read_capabilities`.

aljoscha replied:
pushed a commit that does this. h/t for finding this!

bkirwi commented on src/storage-client/src/controller.rs:
```diff
@@ -1299,30 +1357,58 @@ where
             if desc.sink.from != *from_id {
                 return Err(StorageError::SourceIdReused(*id));
```
I know this isn't you, but this seems more like an input inconsistency than ID reuse... am I understanding that right?

bkirwi commented on src/storage-client/src/controller.rs:
```diff
@@ -1505,38 +1601,53 @@ where
-                panic!("Reference to absent collection");
-            }
-        }
+            if let Ok(collection) = self.collection_mut(*id) {
```
So this is almost entirely besides the point of this PR, but -

For a few reasons (the "what to do with sink errors" question, cluster unification, etc.) I've been wondering whether it would make sense for a sink to have an output collection. It would be a slightly odd collection, in that it would never have any actual records. But it would give us somewhere to report errors - either just passing along the errors from the sink's input, or additional errors like an encoding failure. And it would help make code like this more uniform... we wouldn't need distinct paths for regular collections and exports.

I haven't actually worked through the details, so, grain of salt. Just a little galaxy brain for a Friday...

aljoscha replied:
Yeah, this is an interesting idea. The fact that we silently ignore errors from the input `err` collection has been bugging me a fair bit now... :see_no_evil:

petrosagg replied:
I love this idea! Many things fall into place like querying `EXPLAIN TIMESTAMP FOR SELECT * FROM sink` will tell you how far along that sink is

bkirwi reviewed (commented):
I'm going to need to come back to this after revisiting some of the sink code, but I'm generally very enthusiastic about handling this in a more principled way, and the overall shape seems great.

Leaving a few early comments now, but I'll aim to do a full review by early next week.

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -420,6 +424,33 @@ pub enum ReadPolicy<T> {
 
+impl<T> ReadPolicy<T>
+where
+    T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation,
```
absolutely right!

aljoscha commented:
I pushed a commit that adds a link to where the sink calculates the reported frontier. I also pushed an (unrelated) commit that cleans up the types we use for reporting export errors.

jkosh44 reviewed (approved):
FWIW, `util.rs` LGTM

aljoscha commented:
@bkirwi and @sploiselle I believe I found and fixed the bug that was causing CI to fail: https://github.com/MaterializeInc/materialize/pull/17603/commits/00948b27ec7c386a57d2c8978685e44c46f08f44

This is a bit subtle. Before my change, when downgrading the `implied_capability` of any collection we were taking into account any exports that were using it. This meant that the `implied_capability` was (wrongly, in my opinion) held at what exports need. In my opinion, and this is what underlies the philosophy of this change, the `implied_capability` is what is controlled by the read policy that downgrades the since in accordance with changes in the `upper`, while the `read_capabilities` should be used to track both the requirements of the `implied_capability` _and_ any other requirements (holds installed by the coordinator, and with my change also sinks).

The code in `create_exports` was still working under the old assumption, but I now changed it to use the `acquired_since`, for which we previously installed read holds.

The way the bug could happen before this fix:

1.  - coordinator calls `prepare_sink_export` to install read holds
1.  - time passes (seconds, sometimes, according to log output from tests)
1.  - the `implied_capability` of the input collection gets downgraded because the upper changes. Though, crucially, the overall since of the input collection is _not_ advanced because we still have our read holds installed
1.  - coordinator calls `create_exports`, which uses the (now advanced) `implied_capability` instead of the acquired since, as it should. That `implied_capability` is too far ahead for the purposes of the sink

Context rustdoc for `read_capabilities` and `implied_capability`: https://github.com/MaterializeInc/materialize/blob/60e35938c2988ce5607eaca1f4e3a19ae0e2c88e/src/storage-client/src/controller.rs#L2437

bkirwi commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+    /// Install read capabilities on the given `storage_dependencies` and return
+    /// the frontier at which we are guaranteed to be able to read from all
+    /// dependencies.
+    fn install_read_capabilities(
```
I notice this is called for both `create_collections` and `create_exports`, but its mate `remove_read_capabilities` is only called for sinks. Why's that?

bkirwi commented on src/storage-client/src/controller.rs:
```diff
@@ -1258,83 +1313,151 @@ where
+            //
+            // The write frontier that sinks communicate back to the controller
+            // indicates that all further writes will happen at a time `t` such
+            // that `!timely::ParitalOrder::less_than(&t, &write_frontier)` is
```
```suggestion
            // that `!timely::PartialOrder::less_than(&t, &write_frontier)` is
```

But also there's a type error here... we're comparing a frontier to a timestamp.

aljoscha replied:
I'm reserving changes to this block of comments and how the sink does it's upper reporting to follow-up work. Hoping to get the rest of this merged rather soon to get Sean unblocked.

bkirwi replied:
That seems fine, especially since this comment is simply relocated.

bkirwi commented on src/storage-client/src/controller.rs:
```diff
@@ -2281,19 +2492,37 @@ pub struct ExportState<T> {
+    pub read_capability: Antichain<T>,
+
+    /// The policy to use to downgrade `self.read_capability`.
+    pub read_policy: ReadPolicy<T>,
```
Assuming that we trust the sink to hold back its own reported write frontier, we can get rid of this, yeah? It seems like the one bit that doesn't have an equivalent on `CollectionState`... the `read_policy` there means something else entirely IIUC.

petrosagg replied:
If we subscribe to the idea that sinks are just external materialized views (I do) then handling them as just another collection unifies this handling.

Regardless of this however the messages that are coming back from `clusterd` regarding sink progress should be the pristine write frontier of the sink and if there is a sink specific minus one that we want to do for some reason then this is a controller level concern (just like every other type of object). Also, reporting `F-1` means that `Antichain::from_elem(T::minimum())` has dual meaning, both that nothing is persisted and that timestamp 0 is persisted which is another reason I think we shouldn't manipulate the write frontier

bkirwi reviewed (commented):
Overall I'm quite enthusiastic about the change, but I did leave a couple questions I'd like to settle before I'm totally confident I'm understanding it correctly. (Thanks!)

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+            let dep_collection = self.collection(dep_id.clone())?;
+            for time in capability.iter() {
+                if dep_collection.read_capabilities.count_for(time) <= 0 {
+                    panic!("dependency {dep_id} did not have acquired read capabilities for {:?}). Read capabilities: {:?}", capability, dep_collection.read_capabilities);
```
NIT: this looks like a manual assert
```rust
assert!(
    dep_collection.read_capabilities.count_for(time) > 0,
    "dependency {dep_id} did not have acquired read capabilities for {:?}). Read capabilities: {:?}",
    capability, dep_collection.read_capabilities
)
```

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+            changes.update(time.clone(), -1);
+        }
+
+        // TODO: This seems involved, but we also don't do it too frequently.
```
What is the `TODO` to be done here? Seems fine to me as-is

aljoscha replied:
I'll remove it, but had it there in case there were concerns about the checking logic.

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+
+        // TODO: This seems involved, but we also don't do it too frequently.
+        for dep_id in storage_dependencies {
+            let dep_collection = self.collection(dep_id.clone())?;
```
Shouldn't this check be in `update_read_capabilities`? Ensuring that we never end up with negative accumulations of timestamps?

aljoscha replied:
it's another one of the places where `update_read_capabilities` is very close to how compute does it, and it also doesn't check the updates.

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+        if let DataSource::Ingestion(ingestion) = &collection_description.data_source {
+            match &ingestion.desc.envelope {
+                SourceEnvelope::Debezium(envelope_debezium) => {
+                    let tx_metadata_topic = envelope_debezium.dedup.tx_metadata.clone();
```
You should be able to avoid the clone: `let tx_metadata_topic = &envelope_debezium.dedup.tx_metadata;`

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+    ) -> Vec<GlobalId> {
+        let mut result = Vec::new();
+
+        if let DataSource::Ingestion(ingestion) = &collection_description.data_source {
```
This should also be an exhaustive `match` to future proof it

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1403,14 +1526,11 @@ where
+            let read_capability = export.read_capability.clone();
+            let storage_dependencies = export.storage_dependencies.clone();
+            // TODO: Better error handling/validation?
+            let _ = self.remove_read_capabilities(read_capability, &storage_dependencies);
```
I would either `.unwrap()` this or thread `Result` all the way up. Ignoring it is stressful

aljoscha replied:
remove the `Result` data type. The verification logic is asserting, I think problems with this always point at a bug. wdyt?

petrosagg replied:
I saw the assertion but there is also the fetching of the collection by id that had a `?`. Up to you, I'm fine with panicking. I think we would panic anyway somewhere up the stack if the id was missing

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1486,17 +1606,30 @@ where
+
+            let mut new_read_capability = policy.frontier(collection.write_frontier.borrow());
+
+            if timely::order::PartialOrder::less_equal(
```
Also, should this be `less_than`? If they are equal there is nothing to do

aljoscha replied:
This is how the compute controller does it as well, and I wanted to keep those as similar as possible:
https://github.com/MaterializeInc/materialize/blob/594bd9809f17979fe3768a7caa3fd5bf5e804f44/src/compute-client/src/controller/instance.rs#L799

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1486,17 +1606,30 @@ where
     fn set_read_policy(&mut self, policies: Vec<(GlobalId, ReadPolicy<Self::Timestamp>)>) {
         let mut read_capability_changes = BTreeMap::default();
+
         for (id, policy) in policies.into_iter() {
```
not your code, but `into_iter()` is always implied in for loops, it can be `for (id, policy) in policies {`

petrosagg commented on src/storage-client/src/controller.rs:
```diff
@@ -1505,38 +1638,53 @@ where
+                    }
+                }
+            } else if let Ok(export) = self.export_mut(*id) {
+                if PartialOrder::less_than(&export.write_frontier, new_upper) {
```
I agree. Something like `refresh_policies` sounds right?

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -420,6 +424,33 @@ pub enum ReadPolicy<T> {
+{
+    /// Creates a read policy that lags the write frontier "by one".
+    pub fn step_back() -> Self {
+        Self::LagWriteFrontier(Arc::new(move |upper| {
```
What was the method we landed on in the end, again?

petrosagg replied:
You have two variables `since` and `next_since` as your state. The initial value for both is the current since of the collection. 

For each observed `upper` if `next_since.iter().all(|t| !upper.less_equal(t))) == true`:
    1. `since = next_since`
    2. `next_since = upper`

In all cases you return `since` as the result of the policy

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -1258,83 +1313,151 @@ where
         // Validate first, to avoid corrupting state.
         let mut dedup_hashmap = BTreeMap::<&_, &_>::new();
-        for (CreateExportToken { id, from_id }, desc) in exports.iter() {
+        for (
+            CreateExportToken {
+                id,
+                from_id,
+                acquired_since: _,
+            },
+            desc,
+        ) in exports.iter()
+        {
```
yeah, this was the previous code, but I like yours more!

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -1058,7 +1112,7 @@ where
         let to_register: Vec<_> = futures::stream::iter(enriched_with_metdata)
             .map(|data: Result<_, anyhow::Error>| async move {
-                let (id, description, metadata) = data?;
+                let (id, description, metadata, storage_dependencies, _dependency_since) = data?;
```
because there's a bug! :see_no_evil:

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -1785,9 +1951,109 @@ where
+    /// `collection_description` depends on.
+    ///
+    /// TODO: @sean: This is where the remap shard would slot it.
+    fn get_collection_dependencies(
```
it was name like this because I had a `get_export_dependencies` at one point. I'm renaming and moving the method to `CollectionDescription`.

aljoscha commented:
@bkirwi & @petrosagg I tried to address most of your comments (in mostly separate commits). The bigger outstanding changes, which I would like to punt to follow-up changes are:
 - I still kept the `step_back` `ReadPolicy` instead of the new thing we discussed with Frank. I feel like this has to be done when we also fix how sinks report their upper.
 - I did not change how sinks report their upper, nor did I try and fix the big comment block.
 - I did not try and remove the duplication around `refresh_policies`. There are subtle differences in the fields that `Collections`/`Exports` have. I think we could do sth with a common trait here, but didn't want to spend too much time on it because I want to merge this soon, to unblock @sploiselle. Without piling on (too much) more tech-debt, of course!

bkirwi reviewed (approved):
Thanks for the followup!

A bit puzzled by the assertion failures, but assuming we either discover that these assertions would have failed even before these changes, or we get them fixed up without much structural change, happy to go ahead with this! The bullet points you called out all seem fine to take as a followup.

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -1383,6 +1538,17 @@ where
     }
 
     fn drop_sources_unvalidated(&mut self, identifiers: Vec<GlobalId>) {
+        for id in identifiers.iter() {
```
I think this should also remove the dependencies from the collection when it drops the read capabilities, e.g. something like:

```diff
diff --git a/src/storage-client/src/controller.rs b/src/storage-client/src/controller.rs
index 2076d7971..d1681746c 100644
--- a/src/storage-client/src/controller.rs
+++ b/src/storage-client/src/controller.rs
@@ -1614,13 +1614,13 @@ where
 
     fn drop_sources_unvalidated(&mut self, identifiers: Vec<GlobalId>) {
         for id in identifiers.iter() {
-            let collection = match self.collection(id.clone()) {
+            let collection = match self.collection_mut(*id) {
                 Ok(collection) => collection,
                 Err(_) => continue,
             };
 
             let read_capability = collection.read_capabilities.frontier().to_owned();
-            let storage_dependencies = collection.storage_dependencies.clone();
+            let storage_dependencies = collection.storage_dependencies.drain(..).collect_vec();
             self.remove_read_capabilities(read_capability, &storage_dependencies);
         }
```

because the calls to `remove_read_capabilities` and `set_read_policy` both call `update_read_capabilities`. 
- `remove_read_capabilities`removes the read capability from the dependency
-  `update_read_capabilities` tries to propagate the change to this ID's since to the dependency, even though we've just dropped the capability it holds.

This will result in a double-retraction and panic in in the new `assert` you added [here](https://github.com/MaterializeInc/materialize/pull/17603/files#diff-3154d32c4d0818b6513328628453395000c38bd1238bd1fa9e96b796201b38a6R1751-R1757).

aljoscha commented:
thank you all for the reviews! 🙇
